### ISSUE TITLE
[2.2.1 - scala] SimpleResult#withHeaders ignore duplicate headers

### ISSUE THREAD

**Issue by gervaisb** [playframework/playframework#3544]
Hello,

For my application, I have to accept both Basic and Token authentication. However, it seems that play override duplicates headers.

Given the [HTTP RFC §14.47](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.47), a response can contains multiple WWW-Authenticate headers :

> User agents are advised to take special care in parsing the WWW- Authenticate field value as it might contain more than one challenge, or if more than one WWW-Authenticate header field is provided, the contents of a challenge itself can contain a comma-separated list of authentication parameters.

So I give two WWW-Authenticate headers to [SimpleResult#withHeaders((String, String)*)](https://www.playframework.com/documentation/2.2.1/api/scala/play/api/mvc/SimpleResult.html#withHeaders%28%28String, String%29*%29:SimpleResult[A]) :

```
    Unauthorized(msg).withHeaders(
      (WWW_AUTHENTICATE -> "Basic test"),
      (WWW_AUTHENTICATE -> "Token test") 
    )
```

But the response contains only the last one :

> HTTP/1.1 401 Unauthorized
> Content-Length: 54
> Content-Type: text/plain; charset=utf-8
> WWW-Authenticate: Token test

Changing the headers order prove that the last value override the previous one.

**Comment by crooy** [comment]
So it took me some reading, but it appears that concatenating or duplicating repeated entries are both valid solutions. Now the problem I have with the one currently chosen is that the nghttp2 proxy I want to use does not support reading concatenated headers. 

Could there be any way we switch to duplicated header entries?

**Comment by disaacson** [comment]
+1 for supporting the HTTP feature of multiple headers with the same name. Multiple Vary headers is a valid part of the protocol, but I'm having to write some tortuous code to concatenate values every time I need to add a new Vary header in my request processing pipeline.

**Comment by gmethvin** [comment]
While it is a valid part of the spec for some headers, I don't think it'd be as nice an API if we were to just change the headers map to `Map[String, Seq[String]]`. Allowing multiple of any header by default could make it easy to create invalid responses (e.g. multiple CORS headers), which IMO could be a worse situation than what we have now. The current API is less flexible but makes it hard to shoot yourself in the foot.

We need something that only makes multiple headers/values work for the headers that need it. So I think it would make more sense to do this as part of #5811.

Fortunately, the `Vary` header (like many other headers) supports a comma-separated list of values in lieu of multiple headers, so the problem is easy to work around. You could even add your own extension methods to `Result` to handle headers that support a comma-separated list:

```scala
implicit class ResultMultipleHeaderOps(result: Result) extends AnyVal {
  def withHeaderValues(header: String, values: String*): Result = {
    result.withHeaders(header -> (getHeaderValues(header) ++ values).distinct.mkString(", "))
  }
  def getHeaderValues(header: String): Seq[String] = {
    result.header.headers.get(header).toSeq.flatMap(_.split("""\s*,\s*"""))
  }
}
```

**Comment by disaacson** [comment]
Unfortunately we have multiple Play filters that each need to add a value
to the Vary header. That means we need to use the existing API and then
hack it by concatenating values. It is a poor separation of concerns as the
operations are logically independent but they are forced to know about each
other due to the HTTP response library that has less than ideal fidelity to
the HTTP protocol.

**Comment by disaacson** [comment]
Anther good point to remember is that the HTTP header set is by design
extensible. It would be poor to limit multiple headers with the same name
to a whitelist as that would limit the designed extensibility of the
protocol.

**Comment by gmethvin** [comment]
@disaacson You're right that Play's API doesn't support the full HTTP spec. I'm just not sure if we can reasonably fix it in stable releases without making major changes to the API, and if it's worth it given how rarely it is actually an issue. If you have suggestions on the new API I'd be happy to hear them though.

My point in pointing to #5811 was that we realize the current headers representation is broken, and I'd like to get input on the new API, which hopefully we'll be able to add in 2.6.

**Comment by disaacson** [comment]
That makes sense about the next major release. I see it is tagged with 3.0
instead of 2.6 but I'm not super familiar with Play release plans. Thanks
for considering it!

**Comment by gmethvin** [comment]
@disaacson I re-tagged it as 2.6 since I think we should try to implement it for 2.6 if possible.

**Comment by disaacson** [comment]
Awesome. Thanks!

Maybe both the response and request header objects (using HTTP RFC
terminology not Play terminology) could derive from a common MessageHeader
type to avoid similar situations to what we have where the request supports
multiple headers with same name but the response does not.

**Comment by pauldraper** [comment]
> Irrelevant, the CORS spec says that multiple Access-Control-Allow-Origin headers is illegal
> Rather, the header, according to the spec, may contain an origin list or null.

No, the *Origin* header is a list. Access-Control-Allow-Origin is a single value.

Comma-separated values and multiple headers are in fact, precisely the same. RFC 2616:

> Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma

Some headers are allowed to multi-value (Vary), and some are not (Content-Length). But there is no distinction between multiple headers or single header comma-separated. 

(Except for RFC 6265 which contradicts RFC 2109 and RFC 2616 and establishes Set-Cookie as a Special Snowflake.)

**Comment by pauldraper** [comment]
#5811 is a interesting idea. Also, ambitious.

But regardless of the status of #5811, both `RequestHeader` and `ResponseHeader` should be using the same header API.

Today, `RequestHeader` uses `Headers`, and `ResponseHeader` uses `Map[String, String]`. 

The former respects case-insensitivity, allows multi-valued headers, and is, in general, workable. The later is a naive attempt at HTTP.

`RequestHeader` and `ResponseHeader` should be consistent in using `Headers`.

**Comment by pauldraper** [comment]
On that note, an ugly, but technically correct workaround is to use different cases.

```scala
Unauthorized(msg).withHeaders(
  ("WWW-Authenticate" -> "Basic test"),
  ("www-Authenticate -> "Token test"),
  ("wWw-AuThEnTiCaTe -> "Other test"),
)
```

**Comment by dwickern** [comment]
The workaround with different string casing doesn't work, because the headers are placed into a case-insensitive hash. Please fix this!

**Comment by dwickern** [comment]
A comma-delimited header value *should* be equivalent to the same header repeated, but in practice it's not, at least for `WWW-Authenticate`. Compare browser support for [comma-delimited](https://greenbytes.de/tech/tc/httpauth/#multibasicunknown2) vs [multiple headers](https://greenbytes.de/tech/tc/httpauth/#multibasicunknown2mf) and related bug reports for [firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=669675) and [chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=103220).

Would you guys accept a PR to add a special case for `WWW-Authenticate`, similar to `Set-Cookie`?

**Comment by akkie** [comment]
Any news here? In my opinion, the Play API is inconsistent in this case, because Play can handle duplicate request headers. It should also support duplicate response headers.

**Comment by RaviTejaKomma** [comment]
Any update on this? is this fixed in Play3 with Type-Safe HTTP headers?
